private/pkg/storage/storagegit/bucket.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagegit
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+
+	"github.com/bufbuild/buf/private/pkg/git"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storageutil"
+)
+
+type bucket struct {
+	objectReader git.ObjectReader
+	root         git.Tree
+}
+
+func newBucket(
+	objectReader git.ObjectReader,
+	root git.Tree,
+) (storage.ReadBucket, error) {
+	return &bucket{
+		objectReader: objectReader,
+		root:         root,
+	}, nil
+}
+
+func (b *bucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	entry, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch entry.Mode() {
+	case git.ModeFile, git.ModeExe:
+		data, err := b.objectReader.Blob(entry.Hash())
+		if err != nil {
+			return nil, err
+		}
+		return &namedReader{
+			info:   b.newObjectInfo(path),
+			reader: bytes.NewReader(data),
+		}, nil
+	case git.ModeSymlink:
+		// Symlinks are stored as blobs that reference the target path as a relative
```

Comment:
We don't always want to follow symlinks, its a security issue. For local storage, we have a system of options that feed down from the CLI  entry point, where we do want to follow symlinks, but are not enabled in core. We should evaluate what we want to do here.

Reply:
Yes, that was an open question I meant to ask in this PR. I think I will add back the symlink options.